### ISSUE TITLE
Should error when unwrapping an unwrappable variable

### ISSUE THREAD

**Issue by tinganho** [Microsoft/TypeScript#8453]
`s` cannot be null or undefined below. So I think it is appropriate to error when trying to unwrap.
```ts
let s: string;
id = '';
if (s!) { // should throw an error
    
}
```

**Comment by basarat** [comment]
I think you meant : 

```ts
let s: string;
s = '';
if (!s) { // should throw an error

}
```

That said it can cause issues as people might try defensive coding due to external influences e.g.  https://github.com/Microsoft/TypeScript/pull/8452 now allows people to do `null/undefined` checks even though they are not in the domain of the type :rose:

**Comment by tinganho** [comment]
@basarat I actually meant `s!`.

**Comment by basarat** [comment]
> @basarat I actually meant s!.

Isn't it a syntax error?

![image](https://cloud.githubusercontent.com/assets/874898/15004931/d13201e2-1200-11e6-878d-bf919656629e.png)

**Comment by tinganho** [comment]
I'm not getting any error. The unwrap operator removes 'undefined | null' from the type:

<img width="310" alt="screen shot 2016-05-04 at 12 11 11" src="https://cloud.githubusercontent.com/assets/1324476/15005015/52e1121a-11f1-11e6-90ff-f3ac286ec1c7.png">

**Comment by basarat** [comment]
Indeed not an error on master

![image](https://cloud.githubusercontent.com/assets/874898/15005135/d20dc9cc-1203-11e6-9b40-c56c7846ed9c.png)

Apparently its a `NonNullExpression` (like you said, to remove `undefined | null` from the type, a TypeScript type system thing) as opposed to a `PrefixUnaryExpression` (a JavaScript thing). I've got some searching to do. Sorry for misunderstanding :rose:

**Comment by basarat** [comment]
Got it. I really should have read https://github.com/Microsoft/TypeScript/pull/7140 better instead of winging it :rose:  

![image](https://cloud.githubusercontent.com/assets/874898/15005220/ba91fb3c-1204-11e6-8a51-93be4d9bd531.png)

Did spend some time searching for "unwrap operator" :)

**Comment by tinganho** [comment]
> Did spend some time searching for "unwrap operator" :) 

Sorry might have used the wrong terminology.  Non-null assertion operator is more correct.

**Comment by RyanCavanaugh** [comment]
We allow benign coercions everywhere else (e.g. unary `+` is valid on `number`, you can type-assert to the same type, etc) and it doesn't seem to be much of a problem. What's the reasoning to disallow this one?

**Comment by tinganho** [comment]
@RyanCavanaugh I would like if TS did the same with the two example you mentioned. Though I think the unwrap operator is different. It is more common. Actually I found out that had a case where a variable where most of the time of type `T` 99% of the cases and 1% of the cases it was `undefined`. So spontaneously, I put it as `T | undefined`. Though I discovered I needed to put `!` everywhere. So I then switched it to type it as `T` instead. During refactor it could really help me if there was errors where ever I put an unnecessary `!`.

Besides refactoring benefits, I think it is correct to error just because it is unnecessary operation.

Swift also errors:
<img width="445" alt="screen shot 2016-05-05 at 11 22 38" src="https://cloud.githubusercontent.com/assets/1324476/15035431/7b4312a8-12b4-11e6-89d7-b0bd02192a3a.png">

**Comment by mhegazy** [comment]
what about type parameters?

**Comment by tinganho** [comment]
>what about type parameters?

Can you elaborate?

**Comment by mhegazy** [comment]
```ts
function f<T>(a: T) {
    return a!; // is this an error or not?
}
```

**Comment by tinganho** [comment]
It should error. I think a nullable type argument should be declared at the parameter declaration, because it is more declarative and it is not the call site that decides if it is nullable or not. 

A nullable type passed in a non-nullable type argument function should also error.

```ts
function f<T>(a: T) {
    return a!; // is this an error or not? yes it is
}

let str: string | undefined;
f(str); //error str is nullable

function f<T>(a: T | undefined) {
    return a!; // is this an error or not? no it isn't
}
```

FWIW, that is what swift does. 

<img width="400" alt="screen shot 2016-05-07 at 10 53 26" src="https://cloud.githubusercontent.com/assets/1324476/15089903/fe41d6d8-1441-11e6-8a61-71ab8242a035.png">

(Interestingly Swift allows nullable passed as an non-null type argument):

<img width="335" alt="screen shot 2016-05-07 at 10 54 57" src="https://cloud.githubusercontent.com/assets/1324476/15089910/3b7439d8-1442-11e6-8f9c-9d9c2f91bfbb.png">

**Comment by tinganho** [comment]
Maybe it is best to expose this feature behind a flag. As CFA improves it will break code that it didn't cover before. In my opinon it is good to error, because devs can learn new CFA:s.

**Comment by DanielRosenwasser** [comment]
I think at one point @alexeagle mentioned that if your dependencies weren't updated to use non-null assertions, you could add the assertions as something of a TODO until your dependencies *were* updated (and if I'm misquoting, feel free to clarify).

So it seems like using the assertion is okay in some contexts where you're migrating code, but eventually I think you'd want to get rid of any places where you don't really need those assertions. We should consider how often this scenario pops up.

**Comment by RyanCavanaugh** [comment]
We're going to leave this as an OK thing to do. In addition to precedence with other allowed facile assertions, the real risk here is that it could make *improvements* to the control flow analysis into breaking changes, which would be really unfortunate.
